### ISSUE TITLE
mrview: Behavior of 2D rendering with off-voxel-centre focus point

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#663]
Consider the scenario where you load an image into `mrview`, load the _same image_ as an overlay volume, and enter ortho view. Leave interpolation on for both main and overlay images. Now move the focus point so that it no longer lies precisely at the centre of a voxel. The image data displayed is different for the two images: one is selecting the appropriate slice and performing 2D interpolation, the other is performing full 3D interpolation. Yet it's the same image data, displayed (supposedly) at the same focus point.

I realize that turning off 'snap to axis' at this point would activate 3D interpolation of the main image and then they'd look the same. But that requires a deeper level of understanding of what the viewer's doing. Fundamentally, for the main image in this case the image data being displayed is not faithful to the focus point. And it's been confusing enough for a user to raise as a 'bug'.

Two (potentially not mutually exclusive) possibilities:

* Have two levels of viewer 'snapping': the current 'snap-to-axis', which controls the orientation of the camera based on the main image; and a new mode, 'snap-to-voxel', which also forces the focus point to lie at a voxel centre. This prevents sub-voxel offsets in/out of the screen from arising, so the 2D texture rendering remains faithful for all three projections.

* Modify the basic 2D image texture renderer to acquire 2 slices worth of image data, and perform interpolation between the two based on the sub-voxel depth offset.

Open to opinions / suggestions.

**Comment by steso** [comment]
Maybe the snap-to-voxel (center) is too restrictive / only suitable if interpolation is turned off completely.
I'd like to add a third possibility here, as long as interpolation & snap-to-axis is active: 
- In the ortho view, the image volume is really snapping to axis and slice, but the overlay is only snapping to axis and not slice (relative to the image reference frame). The 2D interpolation can easily cope with in-plane subvoxel shifts, so I propose to display the 3D interpolation of the overlay volume at the exact through-plane location of the displayed image slice for each view.

In my opinion, loading 2 slices of the image-volume and perform "pseudo" 3D interpolation to cope with sub-voxel through-plane offsets is the most elegant solution, the other two are just "workarounds" to assure consistency when snap-to-axis is enabled between the image and the overlay volumes for either interpolation turned off (1) or interpolation turned on (3)

**Comment by jdtournier** [comment]
Yes, I can see where that might be confusing for users. I've been thinking about your suggestion, and I can't think of a cleaner way to fix this without losing MRView's 'limited load' ability (which allows fast display from slow connections/storage). It means twice the load time per slice, but realistically only for the first slice displayed - after that, moving one slice up or down will re-use the data already cached by the OS in the previous display operation. It's actually relatively trivial to implement, since the texture for the 2D render is already a 3D texture with depth 1 (it was easier to keep all the render calls consistent that way). So it should just be a matter of making it depth two, ensuring all the load ops fill both slices appropriately, and setting the `stp` texture coordinates appropriately in the shader. I think that would work equally well for linear vs. nearest-neighbour interpolation (although loading the second slice is clearly redundant in the nearest-neighbour case). More than happy for someone to take that on.

The other idea of snapping to voxel centres might also be a good idea, especially if for example users want to ensure their overlays are _exactly_ lined up with their corresponding slice. This wouldn't be an issue with the above approach if both the main and overlay images are rendered using nearest-neighbour interpolation, but not if the overlay is interpolated. I was thinking we could force lock to voxel centres if interpolation is nearest-neighbour (I think this is what @steso was suggesting?), but there's a few use cases I can think of where I'd like to be able to position the focus with subvoxel precision on a nearest-neighbour display (e.g. checking exactly where a streamline terminates relative to the voxel grid, whether an interpolated overlay corresponds to the main image grid, etc). But it's also a relatively uncommon use case, which can be effectively handled by turning off 'lock to axes'. So maybe it's worth thinking about this. It's not actually such a big change in terms of user experience, since the viewer starts with linear interpolation on anyway - it's only when people switch to nearest-neighbour that they'd notice any difference.

As to how to implement this cleanly without breaking user interaction, it might not be as easy as it sounds. The simplest would be to round the _focus_ to the nearest voxel centre, but that will probably make slice movements through plane really clunky, especially with devices that allow very fine control (e.g. MacOSX pointing devices), since each update is too small by itself to move the focus over to the next slice, so it keeps rounding back to the current slice, without ever scrolling through. We may need to introduce another position vector, in addition to _focus_ and _target_, which holds the genuine focus position, with the current _focus_ set to the rounded version of the genuine focus when nearest-neighbour interpolation and 'lock to axes' are on, and a straight copy of the genuine focus otherwise. All mouse / keyboard interactions would modify the genuine focus, and existing render code that relies on _focus_ would hopefully remain largely unaffected. 

Thoughts...?
